mpiInfo: Linking Boost C++11

`mpiInfo` should be compiled as C++11 so the "standard C++11 ABI boost" we use for PIConGPU can also be used for it without further headaches.

Currently it fails with
```
[81i61029@messi build]$ make -j install
Scanning dependencies of target picongpu-hostonly
[  5%] Building NVCC (Device) object build_cuda_memtest/CMakeFiles/cuda_memtest.dir/cuda_memtest_generated_cuda_memtest.cu.o
[ 11%] Building NVCC (Device) object build_cuda_memtest/CMakeFiles/cuda_memtest.dir/cuda_memtest_generated_tests.cu.o
Scanning dependencies of target mpiInfo
nvcc warning : The 'compute_20', 'sm_20', and 'sm_21' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
nvcc warning : The 'compute_20', 'sm_20', and 'sm_21' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
[ 17%] Building CXX object build_mpiInfo/CMakeFiles/mpiInfo.dir/main.cpp.o
[ 23%] Building CXX object CMakeFiles/picongpu-hostonly.dir/include/plugins/common/particlePatches.cpp.o
[ 29%] Building CXX object CMakeFiles/picongpu-hostonly.dir/include/ArgsParser.cpp.o
[ 35%] Building CXX object CMakeFiles/picongpu-hostonly.dir/include/plugins/common/stringHelpers.cpp.o
[ 41%] Building CXX object CMakeFiles/picongpu-hostonly.dir/include/plugins/hdf5/openPMD/patchReader.cpp.o
nvcc warning : The 'compute_20', 'sm_20', and 'sm_21' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
nvcc warning : The 'compute_20', 'sm_20', and 'sm_21' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
[ 47%] Linking CXX executable mpiInfo
CMakeFiles/mpiInfo.dir/main.cpp.o: In function `boost::program_options::typed_value<bool, char>::xparse(boost::any&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) const':
main.cpp:(.text._ZNK5boost15program_options11typed_valueIbcE6xparseERNS_3anyERKSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaISB_EE[_ZNK5boost15program_options11typed_valueIbcE6xparseERNS_3anyERKSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaISB_EE]+0x17): undefined reference to `boost::program_options::validate(boost::any&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, bool*, int)'
CMakeFiles/mpiInfo.dir/main.cpp.o: In function `boost::program_options::typed_value<bool, char>::name() const':
main.cpp:(.text._ZNK5boost15program_options11typed_valueIbcE4nameEv[_ZNK5boost15program_options11typed_valueIbcE4nameEv]+0x22): undefined reference to `boost::program_options::arg[abi:cxx11]'
main.cpp:(.text._ZNK5boost15program_options11typed_valueIbcE4nameEv[_ZNK5boost15program_options11typed_valueIbcE4nameEv]+0x28): undefined reference to `boost::program_options::arg[abi:cxx11]'
CMakeFiles/mpiInfo.dir/main.cpp.o: In function `boost::program_options::basic_command_line_parser<char>::extra_parser(boost::function1<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&>)':
main.cpp:(.text._ZN5boost15program_options25basic_command_line_parserIcE12extra_parserENS_9function1ISt4pairINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESA_ERKSA_EE[_ZN5boost15program_options25basic_command_line_parserIcE12extra_parserENS_9function1ISt4pairINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESA_ERKSA_EE]+0x46): undefined reference to `boost::program_options::detail::cmdline::set_additional_parser(boost::function1<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&>)'
CMakeFiles/mpiInfo.dir/main.cpp.o: In function `boost::program_options::basic_command_line_parser<char>::basic_command_line_parser(int, char const* const*)':
main.cpp:(.text._ZN5boost15program_options25basic_command_line_parserIcEC2EiPKPKc[_ZN5boost15program_options25basic_command_line_parserIcEC5EiPKPKc]+0x298): undefined reference to `boost::program_options::to_internal(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
main.cpp:(.text._ZN5boost15program_options25basic_command_line_parserIcEC2EiPKPKc[_ZN5boost15program_options25basic_command_line_parserIcEC5EiPKPKc]+0x327): undefined reference to `boost::program_options::detail::cmdline::cmdline(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&)'
CMakeFiles/mpiInfo.dir/main.cpp.o: In function `main':
main.cpp:(.text.startup+0x5a): undefined reference to `boost::program_options::options_description::options_description(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int, unsigned int)'
CMakeFiles/mpiInfo.dir/main.cpp.o:(.rodata._ZTVN5boost15program_options11typed_valueIbcEE[_ZTVN5boost15program_options11typed_valueIbcEE]+0x40): undefined reference to `boost::program_options::value_semantic_codecvt_helper<char>::parse(boost::any&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, bool) const'
collect2: error: ld returned 1 exit status
make[2]: *** [build_mpiInfo/mpiInfo] Error 1
make[1]: *** [build_mpiInfo/CMakeFiles/mpiInfo.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 52%] Linking CXX static library libpicongpu-hostonly.a
[ 52%] Built target picongpu-hostonly
Scanning dependencies of target cuda_memtest
[ 58%] Building CXX object build_cuda_memtest/CMakeFiles/cuda_memtest.dir/misc.cpp.o
[ 64%] Linking CXX executable cuda_memtest
[ 64%] Built target cuda_memtest
make: *** [all] Error 2
```

Reported in https://github.com/ComputationalRadiationPhysics/picongpu/issues/1984#issuecomment-321246238

--- a/src/mpiInfo/CMakeLists.txt
+++ b/src/mpiInfo/CMakeLists.txt
@@ -45,6 +45,16 @@ list(APPEND CMAKE_PREFIX_PATH "$ENV{CMAKE_PREFIX_PATH}")
 list(APPEND "/usr/lib/x86_64-linux-gnu/")
 
 
+###############################################################################
+# Language Flags
+###############################################################################
+
+# enforce C++11
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+set(CMAKE_CXX_EXTENSIONS OFF)
+set(CMAKE_CXX_STANDARD 11)
+
+
 ################################################################################
 # Warnings
 ################################################################################

--- a/src/tools/png2gas/CMakeLists.txt
+++ b/src/tools/png2gas/CMakeLists.txt
@@ -52,6 +52,16 @@ set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH}
     ${CMAKE_CURRENT_SOURCE_DIR}/../../../thirdParty/cmake-modules/)
 
 
+###############################################################################
+# Language Flags
+###############################################################################
+
+# enforce C++11
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+set(CMAKE_CXX_EXTENSIONS OFF)
+set(CMAKE_CXX_STANDARD 11)
+
+
 ################################################################################
 # Build type (debug, release)
 ################################################################################

--- a/src/tools/png2gas/png2gas.cpp
+++ b/src/tools/png2gas/png2gas.cpp
@@ -125,7 +125,7 @@ int main(int argc, char **argv)
     if (!parseCmdLine(argc, argv, options))
         return -1;
 
-    MPI_Init(NULL, NULL);
+    MPI_Init(nullptr, nullptr);
 
     Dimensions data_size(options.dataSize);
     std::cout << "Creating density data with size " << data_size.toString() << std::endl;

--- a/src/tools/splash2txt/CMakeLists.txt
+++ b/src/tools/splash2txt/CMakeLists.txt
@@ -48,6 +48,16 @@ set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH}
 include_directories(include)
 
 
+###############################################################################
+# Language Flags
+###############################################################################
+
+# enforce C++11
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+set(CMAKE_CXX_EXTENSIONS OFF)
+set(CMAKE_CXX_STANDARD 11)
+
+
 ################################################################################
 # Build type (debug, release, relwithdebinfo)
 ################################################################################

--- a/src/tools/splash2txt/splash2txt.cpp
+++ b/src/tools/splash2txt/splash2txt.cpp
@@ -218,7 +218,7 @@ int main( int argc, char** argv )
         return 1;
     }
 
-    ITools *tools = NULL;
+    ITools *tools = nullptr;
     switch ( options.fileMode)
     {
         case FM_SPLASH: tools = new ToolsSplashParallel( options, mpi_topology, *outStream );

--- a/src/tools/splash2txt/tools_adios_parallel.cpp
+++ b/src/tools/splash2txt/tools_adios_parallel.cpp
@@ -61,7 +61,7 @@ void ToolsAdiosParallel::convertToText()
 
         pVarInfo = adios_inq_var(pFile, nodeName.c_str());
 
-        varTypeSize = adios_type_size(pVarInfo->type, NULL);
+        varTypeSize = adios_type_size(pVarInfo->type, nullptr);
 
         // get number of elements combined in a dataset
         for(int j = 0; j < pVarInfo->ndim; j++)
@@ -71,7 +71,7 @@ void ToolsAdiosParallel::convertToText()
         // allocate memory
         P = (uint8_t*) malloc (sizeof(uint8_t) * varTypeSize * varElement);
 
-        adios_schedule_read(pFile, NULL, nodeName.c_str(), 0, 1, P);
+        adios_schedule_read(pFile, nullptr, nodeName.c_str(), 0, 1, P);
 
         adios_perform_reads(pFile, 1);
 

--- a/src/tools/splash2txt/tools_splash_parallel.cpp
+++ b/src/tools/splash2txt/tools_splash_parallel.cpp
@@ -132,7 +132,7 @@ void ToolsSplashParallel::printParticles(std::vector<ExDataContainer> fileData)
                 }
 
                 void* element = container->getElement(i);
-                assert(element != NULL);
+                assert(element != nullptr);
 
                 printElement(data_type, element, iter->unit, m_options.delimiter);
                 numElementsProcessed[container]++;
@@ -196,7 +196,7 @@ void ToolsSplashParallel::printFields(std::vector<ExDataContainer> fileData)
                             iter->container->getIndex(0)->getDataType();
 
                     void* element = iter->container->getElement(index);
-                    assert(element != NULL);
+                    assert(element != nullptr);
 
                     printElement(data_type, element, iter->unit, m_options.delimiter);
                 }
@@ -233,7 +233,7 @@ void ToolsSplashParallel::convertToText()
     try
     {
         dc.readAttribute(m_options.step, m_options.data[0].c_str(), DOMCOL_ATTR_CLASS,
-                &ref_data_class, NULL);
+                &ref_data_class, nullptr);
     } catch (const DCException&)
     {
         errorStream << "Error: No domain information for dataset '" << m_options.data[0] << "' available." << std::endl;
@@ -266,7 +266,7 @@ void ToolsSplashParallel::convertToText()
 
         DomainCollector::DomDataClass data_class = DomainCollector::UndefinedType;
         dc.readAttribute(m_options.step, iter->c_str(), DOMCOL_ATTR_CLASS,
-                &data_class, NULL);
+                &data_class, nullptr);
         if (data_class != ref_data_class)
             throw std::runtime_error("All requested datasets must be of the same data class");
 
@@ -282,7 +282,7 @@ void ToolsSplashParallel::convertToText()
             // poly type
 
             excontainer.container = dc.readDomain(m_options.step, iter->c_str(),
-                    Domain(total_domain.getOffset(), total_domain.getSize()), NULL, true);
+                    Domain(total_domain.getOffset(), total_domain.getSize()), nullptr, true);
         } else
         {
             // grid type
@@ -302,7 +302,7 @@ void ToolsSplashParallel::convertToText()
                 }
 
             excontainer.container = dc.readDomain(m_options.step, iter->c_str(),
-                    Domain(offset, domain_size), NULL);
+                    Domain(offset, domain_size), nullptr);
         }
 
         // read unit
@@ -312,7 +312,7 @@ void ToolsSplashParallel::convertToText()
             try
             {
                 dc.readAttribute(m_options.step, iter->c_str(), "unitSI",
-                        &(excontainer.unit), NULL);
+                        &(excontainer.unit), nullptr);
             } catch (const DCException&)
             {
                 if (m_options.verbose)
@@ -333,7 +333,7 @@ void ToolsSplashParallel::convertToText()
         file_data.push_back(excontainer);
     }
 
-    assert(file_data[0].container->get(0)->getData() != NULL);
+    assert(file_data[0].container->get(0)->getData() != nullptr);
 
     // write to file
     //
@@ -391,7 +391,7 @@ void ToolsSplashParallel::listAvailableDatasets()
 {
     // number of timesteps in this file
     size_t num_entries = 0;
-    dc.getEntryIDs(NULL, &num_entries);
+    dc.getEntryIDs(nullptr, &num_entries);
     if (num_entries == 0)
     {
         m_outStream << "no entries in file" << std::endl;
@@ -399,7 +399,7 @@ void ToolsSplashParallel::listAvailableDatasets()
     }
 
     int32_t *entries = new int32_t[num_entries];
-    dc.getEntryIDs(entries, NULL);
+    dc.getEntryIDs(entries, nullptr);
     std::sort(entries, entries + num_entries);
 
     m_outStream << std::endl
@@ -418,9 +418,9 @@ void ToolsSplashParallel::listAvailableDatasets()
     // available data sets in this file
     std::vector<DataCollector::DCEntry> dataTypeNames;
     size_t numDataTypes = 0;
-    dc.getEntriesForID(entries[0], NULL, &numDataTypes);
+    dc.getEntriesForID(entries[0], nullptr, &numDataTypes);
     dataTypeNames.resize(numDataTypes);
-    dc.getEntriesForID(entries[0], &(dataTypeNames.front()), NULL);
+    dc.getEntriesForID(entries[0], &(dataTypeNames.front()), nullptr);
 
     // parse dataTypeNames vector in a nice way for stdout
     m_outStream << "Available data field names:";

